onnxruntime/core/framework/execution_provider.cc

```diff
@@ -52,103 +52,64 @@
 #endif
 }
 
-// Returns true if an allocator was found and replaced
-static bool FindAndReplaceAllocator(const OrtMemoryInfo& mem_info,
-                                    const MemoryInfoSet& mem_info_set,
-                                    AllocatorMap& allocators,
-                                    AllocatorPtr replacing_allocator) {
-  auto ite = mem_info_set.find(mem_info);
-
-  if (ite != mem_info_set.end()) {
-    const int key = MakeKey(mem_info.id, mem_info.mem_type);
-    allocators[key] = replacing_allocator;
-    return true;
-  }
-
-  return false;
-}
-
 // Update allocator in the provider if already present; ignore if not.
+// We match using the device id, OrtMemType and OrtDevice info.
+// We ignore the allocator name, and OrtAllocatorType (whether internally an arena is used or not).
+// TODO: We should remove OrtAllocatorType from OrtMemoryInfo as it's an implmentation detail of the allocator.
 void IExecutionProvider::ReplaceAllocator(AllocatorPtr allocator) {
   const auto& info = allocator->Info();
 
-  if (FindAndReplaceAllocator(info, mem_info_set_, allocators_, allocator)) {
-    // We found an allocator corresponding to the provided
-    // allocator's OrtMemoryInfo and we replaced it with the
-    // provided allocator.
-    // We return back.
-    return;
-  }
+  auto iter = allocators_.find(MakeKey(info.id, info.mem_type));
+  if (iter != allocators_.end()) {
+    // check device as mem_type is relative to the device
+    // e.g. OrtMemTypeDefault is CPU for a CPU EP and GPU for a CUDA EP. An individual EP will only have one
+    // allocator for an OrtMemType value, so this check is to ensure we don't replace with an incompatible allocator.
+    if (iter->second->Info().device == info.device) {
+      IAllocator* existing_alloc = iter->second.get();
+      for (auto& entry : allocator_list_) {
+        if (entry.get() == existing_alloc) {
```

Comment:
allocator_list_ wasn't previously updated but should be in sync with the values in allocators_